mpeg2/src/interleaving_muxer.rs

```diff
@@ -0,0 +1,463 @@
+use super::EncodeError;
+use crate::muxer::{Muxer, Packet, Stream, StreamConfig};
+use alloc::vec::Vec;
+use core2::io::Write;
+use std::collections::VecDeque;
+use std::time::Duration;
+
+const TS_33BIT_MASK: u64 = 0x1FFFFFFFF;
+const TS_ROLLOVER_THREAD: u64 = 1 << 20;
+
+pub struct InterleavingMuxer<W: Write> {
+    inner: Muxer<W>,
+    max_buffer_duration_90khz: u64,
+    largest_ts_in_buffer: u64,
+    streams: Vec<InterleavingStream>,
+}
+
+pub struct InterleavingStream {
+    inner: Stream,
+    buffered_packets: VecDeque<Packet>,
+    last_written_ts: u64,
+}
+
+impl<W: Write> InterleavingMuxer<W> {
+    pub fn new(w: W, max_buffer_duration: Duration) -> Self {
+        let muxer = Muxer::new(w);
+        Self {
+            inner: muxer,
+            max_buffer_duration_90khz: (max_buffer_duration.as_millis() * 90) as u64,
+            largest_ts_in_buffer: 0,
+            streams: vec![],
+        }
+    }
+
+    pub fn add_stream(&mut self, config: StreamConfig) {
+        let inner = self.inner.new_stream(config);
+        let interleaving_stream = InterleavingStream {
+            inner,
+            buffered_packets: VecDeque::with_capacity(6),
+            last_written_ts: 0,
+        };
+        self.streams.push(interleaving_stream);
+    }
+
+    #[inline]
+    fn get_most_significant_bit(&self, mut n: u64) -> u64 {
+        while n > 1 && (n & (n - 1) != 0) {
+            n &= n - 1;
+        }
+        n
+    }
+
+    fn handle_ts_rollover(&mut self, stream_index: usize, packet: &mut Packet) {
+        let s = &self.streams[stream_index];
+        let buffered_packets = &s.buffered_packets;
+        if let Some(ref mut dts) = packet.dts_90khz {
+            let largest_ts = buffered_packets
+                .back()
+                .map_or(s.last_written_ts, |p| p.dts_90khz.or(p.pts_90khz).unwrap_or(s.last_written_ts));
```

Comment:
I think this can be simplified:

```suggestion
                .and_then(|p| p.dts_90khz.or(p.pts_90khz)).unwrap_or(s.last_written_ts);
```

Reply:
👍